### ISSUE TITLE
Adding Addenda-record

### ISSUE THREAD

**Issue by Thorium** [ekonbenefits/NachaSharp#6]
How can I add an Addenda-record to PPD-entry?
It seems the [EntryAddenda05](https://github.com/ekonbenefits/NachaSharp/blob/cdee1979b91ba5047c3aae0861876addd2a41c1c/NachaSharp/NachaRecords/EntryDetail/EntryAddenda.fs#L45) would be according to the [specs](https://resource1.personifycorp.com/PersonifyOnlineHelp/7.7.0/mergedProjects/eft/Addenda_Record_Format.htm).
However creating one crashes runtime:
```fsharp
let info = "testing..."
entry.Addenda.Add(NachaSharp.EntryAddenda05(info))
//System.IndexOutOfRangeException: Index was outside the bounds of the array.
//at FSharp.Data.FlatFileMeta.FlatRow.SetColumn[T](T value, FSharpOption`1 memberName)
//at NachaSharp.EntryAddenda05.set_AddendaSeqNum(Int32 value)
```

Then if I try to create it manually, I don't have access to the sequence number needed:

```fsharp
let seqNum = "0000001" // How to get the internal Entry Detail Sequence Number?
entry.Addenda.Add(NachaSharp.EntryAddendaWildCard("705" + info.PadRight(80, ' ') + "0001" + seqNum))
```

**Comment by Thorium** [comment]
This is an example of addenda record:
https://www.treasurysoftware.com/ach/ach-file-addenda-records.aspx

There was no unit-test of row starting with 705. I tried to add it to current unit test transaction files, and the tests started to fail.
So I expect this is an actual bug rather than missing documentation issue.

**Comment by abelbraaksma** [comment]
This appears to be caused by the linked library [`FSharp.Data.FlatFileMeta`](https://github.com/ekonbenefits/FSharp.Data.FlatFileMeta), which is the grandparent of all Addenda and Entry types. When a record is created with initial `rowData`, this data isn't validated. If the string is not the size of the record (here: 94), it will raise an exception as soon as some property is set beyond the size of that string.

This is not just a problem with `EntryAddenda05`, it can be reproduced with any type. In the example I manually set `AllowMutation`, though internally this is always set on the root record, i.e. it just mimics what already happens.

```f#
[<Fact>]
let ``Create Non-blank EntryCCD`` () =
    let control =  EntryCCD("", "6some data")
    control.AllowMutation <- true           // IndexOutOfRangeException, because this triggers this.Changed, which calls a setter
    control.TraceNumber <- "55"
    control.TraceNumber |> should equal "55"
```

This first example fails because setting the `AllowMutation` triggers the `Changed` event, which in turn runs `CalculateImpl`, which tries to set a field that's out of range of the string "6some data":

```f#
this.AddendaRecordedIndicator <- this.Addenda |> Seq.length 
```

The following is a minimal repro of `Addenda05` which fails on the setter. This mimics what happens internally:

```f#
[<Fact>]
let ``Create Non-blank EntryAddenda05`` () =
    let control =  EntryAddenda05("705Power of the Dog")
    control.AllowMutation <- true
    control.RecordTypeCode <- "7"       // fine, within range of the string
    control.EntrySeqNum <- "5"          // error IndexOutOfRangeException
    control.RecordTypeCode |> should equal "7"
```

The offending line is in `FlatFileMeta.fs`, from the linked project:

```f#
    member private this.LazySetup() =
        if columnMap.Count = 0 then
            let totalLength, orderedKeys, mapMeta = this.Setup()
      
            columnLength <- totalLength
            columnKeys <- orderedKeys
            rawData <- match rowInput with
                        | Some (row) -> 
                            // this is NOT where the error is raised, but what causes it down the line
                            row |> Array.ofSeq |> Array.map string      // this array MUST be of correct size, but usually isn't
                        | None -> Array.init totalLength (fun _ -> " ")
            columnMap <- mapMeta
            this.PostSetup()
```

A trivial fix (which doesn't yet deal with strings that are too long) would be something like this:

```f#
row.PadRight totalLength
|> Array.ofSeq |> Array.map string
```

It's probably also a good idea to have a few tests that actually create some full-blown records, instead of default empty ones.

**Comment by Thorium** [comment]
Thanks Abel,

Reading this, I think there is a possible hack-workaround for this:
When creating a entry, first set it to empty, to trigger the padding not to go 3+80 or whatever:

```fsharp
let info = "testing..."
let addenda = NachaSharp.EntryAddenda05("")
addenda.AllowMutation <- true
addenda.PaymentInfo <- info
addenda.AddendaTypeCode <- 5
entry.AllowMutation <- true
entry.Addenda.Add addenda
```

Now, the entry doesn't have sequence number at that point of time, and the addenda needs the same, so after the entry is attached to a batch `myBatch.Entries.Add entry` then the entry has the sequence number, and we need to still fix the addenda sequence numbers, e.g.:

```fsharp
entry.Addenda |> Seq.iter(
    function
    | :? NachaSharp.EntryAddenda05 as a -> a.EntrySeqNum <- e.TraceNumber.Substring(e.TraceNumber.Length - 7)
    | _ -> ()
)
```

**Comment by abelbraaksma** [comment]
@Thorium Yes, that works. Likewise, and arguably clearer, is `EntryAddenda05 null`, but that's a matter of taste.

It works because of this line in `FlatFileMeta.fs` in the linked lib, which sets `rowInput` to None, which in turn forces initialization later on:

```f#
let rowInput = Helper.optionOfStringEmpty rowData     // returns None if empty string
```

**Comment by Thorium** [comment]
Great, thanks!
Could we have a NuGet package of the fixed version, please?

**Comment by abelbraaksma** [comment]
@Thorium, looks like a new package was released with version 0.8.2-alpha.30: https://www.nuget.org/packages/NachaSharp/.
